src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
+            var replacementIndices = new ValueListBuilder<int>(initialSpan);
+
+            unsafe
+            {
+                fixed (char* pThis = &_firstChar)
+                {
+                    int matchIdx = 0;
+                    int lastPossibleMatchIdx = this.Length - oldValue.Length;
+                    while (matchIdx <= lastPossibleMatchIdx)
+                    {
+                        char* pMatch = pThis + matchIdx;
+                        for (int probeIdx = 0; probeIdx < oldValue.Length; probeIdx++)
+                        {
+                            if (pMatch[probeIdx] != oldValue[probeIdx])
+                            {
+                                goto Next;
+                            }
+                        }
+                        // Found a match for the string. Record the location of the match and skip over the "oldValue."
+                        replacementIndices.Append(matchIdx);
+                        matchIdx += oldValue.Length;
+                        continue;
+
+                    Next:
+                        matchIdx++;
+                    }
+                }
+            }
+
+            if (replacementIndices.Length == 0)
+                return this;
+
+            // String allocation and copying is in separate method to make this method faster for the case where
+            // nothing needs replacing.
+            string dst = ReplaceHelper(oldValue, newValue, replacementIndices.AsReadOnlySpan());
+
+            replacementIndices.Dispose();
+
+            return dst;
+        }
+
+        private string ReplaceHelper(string oldValue, string newValue, ReadOnlySpan<int> indices)
+        {
+            long dstLength = this.Length + ((long)(newValue.Length - oldValue.Length)) * indices.Length;
+            if (dstLength > int.MaxValue)
+                throw new OutOfMemoryException();
+            string dst = FastAllocateString((int)dstLength);
+
+            ReadOnlySpan<char> thisSpan = this.AsReadOnlySpan();
+            Span<char> dstSpan = new Span<char>(ref dst.GetRawStringData(), dst.Length);
+
+            int thisIdx = 0;
+            int dstIdx = 0;
+
+            for (int r = 0; r < indices.Length; r++)
+            {
+                int replacementIdx = indices[r];
+
+                // Copy over the non-matching portion of the original that precedes this occurrence of oldValue.
+                int count = replacementIdx - thisIdx;
+                if (count != 0)
+                {
+                    thisSpan.Slice(thisIdx, count).CopyTo(dstSpan.Slice(dstIdx));
+                    dstIdx += count;
+                }
+                thisIdx = replacementIdx + oldValue.Length;
+
+                // Copy over newValue to replace the oldValue.
+                newValue.AsReadOnlySpan().CopyTo(dstSpan.Slice(dstIdx));
```

Comment:
why not pull out `newValue.AsReadOnlySpan()` into a local like `thisSpan` ? You need it repeatedly.

Reply:
It actually hurts to cache the Span in a local like that. The code became a bit/smaller faster when I have removed caching of `thisSpan` as well.

The problem is that Span is (pointer+length). It burns two registers and causes you to run out of registers quickly. It is better to recompute it each time here.

Reply:
Ah intersting.